pkg/azure/provider/provider.go

```diff
@@ -0,0 +1,201 @@
+	"github.com/gardener/machine-controller-manager/pkg/util/provider/machinecodes/status"
+	"k8s.io/klog/v2"
+
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/access"
+	clienthelpers "github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/access/helpers"
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/provider/helpers"
+	"github.com/gardener/machine-controller-manager-provider-azure/pkg/azure/utils"
+)
+
+const (
+	createMachineOperationLabel    = "create_machine"
+	deleteMachineOperationLabel    = "delete_machine"
+	listMachinesOperationLabel     = "list_machine"
+	getMachineStatusOperationLabel = "get_machine_status"
+	getVolumeIDsOperationLabel     = "get_volume_ids"
+)
+
+// defaultDriver implements provider.Driver interface
+type defaultDriver struct {
+	factory access.Factory
+}
+
+// NewDefaultDriver creates a new instance of an implementation of provider.Driver. This can be mostly used by tests where we also wish to have our own polling intervals.
+func NewDefaultDriver(accessFactory access.Factory) driver.Driver {
+	return defaultDriver{
+		factory: accessFactory,
+	}
+}
+
+func (d defaultDriver) ListMachines(ctx context.Context, req *driver.ListMachinesRequest) (resp *driver.ListMachinesResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, listMachinesOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	vmNames, err := helpers.ExtractVMNamesFromVirtualMachinesAndNICs(ctx, d.factory, connectConfig, providerSpec.ResourceGroup, providerSpec.Tags)
+	if err != nil {
+		err = status.Error(codes.Internal, fmt.Sprintf("Failed to extract VM names from VMs and NICs for resourceGroup: %s, Err: %v", providerSpec.ResourceGroup, err))
+		return
+	}
+	resp = helpers.ConstructMachineListResponse(providerSpec.Location, vmNames)
+	return
+}
+
+func (d defaultDriver) CreateMachine(ctx context.Context, req *driver.CreateMachineRequest) (resp *driver.CreateMachineResponse, err error) {
+	defer instrument.RecordDriverAPIMetric(err, createMachineOperationLabel, time.Now())
+	providerSpec, connectConfig, err := helpers.ExtractProviderSpecAndConnectConfig(req.MachineClass, req.Secret)
+	if err != nil {
+		return
+	}
+	vmName := req.Machine.Name
+	nicName := utils.CreateNICName(vmName)
+
+	imageReference, plan, err := helpers.ProcessVMImageConfiguration(ctx, d.factory, connectConfig, providerSpec, vmName)
+	if err != nil {
+		return
+	}
+	subnet, err := helpers.GetSubnet(ctx, d.factory, connectConfig, providerSpec)
+	if err != nil {
+		return
+	}
+
+	nicID, err := helpers.CreateNICIfNotExists(ctx, d.factory, connectConfig, providerSpec, subnet, nicName)
+	if err != nil {
+		return
+	}
+
+	vm, err := helpers.CreateOrUpdateVM(ctx, d.factory, connectConfig, providerSpec, imageReference, plan, req.Secret, nicID, vmName)
```

Comment:
Should i call this `CreateVM`? I should have renamed `accesshelpers.CreateVirtualMachine` to `accesshelpers.CreateOrUpdateVirtualMachine` actually. The issue is that from the `Driver` API implementation point of view, in both the new code and in the older code we never check if already exists. So consider that the VM already exists and the consumer makes a call to `CreateMachine`, if it was only creating a VM assuming that no machine exists then it would have failed. The reason it never fails is that the azure API that it calls is `CreateOrUpdate`. Also while coding Driver implementation we cannot assume the correctness of the caller.

Reply:
will be renamed, as decided